crates/wasi-experimental-http-wasmtime/src/lib.rs

```diff
@@ -91,88 +114,101 @@ impl HostCalls {
     // TODO (@radu-matei)
     // Fix the clippy warning.
     #[allow(clippy::unnecessary_wraps)]
-    fn close(
-        st: Rc<RefCell<State>>,
-        _caller: Caller<'_>,
-        handle: WasiHttpHandle,
-    ) -> Result<(), HttpError> {
-        st.borrow_mut().responses.remove(&handle);
+    fn close(st: Arc<RwLock<State>>, handle: WasiHttpHandle) -> Result<(), HttpError> {
+        let mut st = st.write()?;
+        st.responses.remove(&handle);
         Ok(())
     }
 
     /// Read `buf_len` bytes from the response of `handle` and
     /// write them into `buf_ptr`.
     fn body_read(
-        st: Rc<RefCell<State>>,
-        caller: Caller<'_>,
+        st: Arc<RwLock<State>>,
+        memory: Memory,
+        mut store: impl AsContextMut,
         handle: WasiHttpHandle,
         buf_ptr: u32,
         buf_len: u32,
         buf_read_ptr: u32,
     ) -> Result<(), HttpError> {
-        let mut st = st.borrow_mut();
-        // Get the current response body.
-        let mut body = &mut st
-            .responses
-            .get_mut(&handle)
-            .ok_or(HttpError::InvalidHandle(handle))?
-            .body;
-        let memory = memory_get(caller)?;
+        let mut st = st.write()?;
```

Comment:
For what it is worth, I would just panic if poison occurs (which is very very rare, in fact, they are thinking of removing it). Particularly in light of this info from the documentation:

> The precise semantics for when a lock is poisoned is documented on each lock, but once a lock is poisoned then all future acquisitions will return this error.

So if it does get poisoned you are SOL

Reply:
Yeah, but this way we get a chance to return an error to the client.

Reply:
I would avoid a panic in the library for a number of reasons, the most important of which being that the runtime could be executing other modules as well.
On top of that, this also gives us a way of returning an error to the client.

Reply:
As long as that terminates things, I think that is great